pkg/proxy/handlers/spacelister.go

```diff
@@ -34,25 +35,59 @@ func NewSpaceLister(app application.Application) *SpaceLister {
 }
 
 func (s *SpaceLister) HandleSpaceListRequest(ctx echo.Context) error {
-	userID, _ := ctx.Get(context.SubKey).(string)
-	username, _ := ctx.Get(context.UsernameKey).(string)
+
+	workspaces, err := s.ListUserWorkspaces(ctx)
+	if err != nil {
+		return errorResponse(ctx, apierrors.NewInternalError(err))
+	}
+
 	workspaceName := ctx.Param("workspace")
 	doGetWorkspace := len(workspaceName) > 0
 
+	if doGetWorkspace { // specific workspace requested
+		if len(workspaces) != 1 {
+			// not found
+			r := schema.GroupResource{Group: "toolchain.dev.openshift.com", Resource: "workspaces"}
+			return errorResponse(ctx, apierrors.NewNotFound(r, workspaceName))
+		}
+		return getWorkspaceResponse(ctx, workspaces[0])
+	}
+
+	return listWorkspaceResponse(ctx, workspaces)
+}
+
+func (s *SpaceLister) ListUserWorkspaces(ctx echo.Context) ([]toolchainv1alpha1.Workspace, error) {
+	userID, _ := ctx.Get(context.SubKey).(string)
+	username, _ := ctx.Get(context.UsernameKey).(string)
```

Comment:
are those inputs from the "users" ? is there any validation required or already in place ?

Reply:
It's based on the token we get from SSO and we have a middleware function in the proxy that is guaranteed to set them or return an error before we ever get here. See https://github.com/codeready-toolchain/registration-service/blob/4b359be48a41f69fb832c82b333b630210cc5d6e/pkg/proxy/proxy.go#L192

Reply:
I see! There is a `extractUserID` function. 

Thanks for the details!